File: c/matching-brackets/matching_brackets.c
#include "matching_brackets.h"

bool is_paired(const char *input) {
  if (input == NULL || strlen(input) == 0)
    return 1;
  int count = 1;
  char *buff = (char *)malloc(strlen(input) / 2 + 1);
  while (*input) {
    if (*input == '{' || *input == '[' || *input == '(') {
      buff[count] = *input;
      count++;
    } else if (*input == '}') {
      if (buff[count - 1] != '{' || count - 1 < 1) {
        free(buff);
        return 0;
      }
      count--;
    } else if (*input == ']') {
      if (buff[count - 1] != '[' || count - 1 < 1) {
        free(buff);
        return 0;
      }
      count--;
    } else if (*input == ')') {
      if (buff[count - 1] != '(' || count - 1 < 1) {
        free(buff);
        return 0;
      }
      count--;
    }
    input++;
  }
  if (count > 1) {
    free(buff);
    return false;
  }
  free(buff);
  return true;
}

File: csharp/matching-brackets/MatchingBrackets.cs
using System.Collections.Generic;

public static class MatchingBrackets
{
    public static bool IsPaired(string input)
    {
        Stack<char> stack = new Stack<char>();
        string brackets = "()[]{}";
        foreach (char b in input)
        {
            if (!brackets.Contains(b)) { continue; }
            if (b == '[' || b == '{' || b == '(')
            {
                stack.Push(b);
                continue;
            }
            if (stack.Count == 0) { return false; }
            char last = stack.Peek();
            if ((b == ']' && last == '[') || (b == ')' && last == '(') || (b == '}' && last == '{'))
            {
                stack.Pop();
                continue;
            }
            return false;
        }
        return stack.Count == 0;
    }
}
